### PR TITLE
fix: eventsender engine uploads on a cadence

### PR DESCRIPTION
**Pull request by nicklasl**

### CODE REVIEW

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java [hunk 1]
```diff
@@ -44,13 +52,34 @@ public void run() {
...
+    if (pendingFlush != null && !pendingFlush.isDone()) {
+      pendingFlush.cancel(false);
+    }
+    if (flushTimeoutMilliseconds > 0) {
+      pendingFlush =
+          executorService.schedule(this::flush, flushTimeoutMilliseconds, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private void flush() {
```
Comment (nicklasl) [inline src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:74]: What happens if this is triggered both from `numberOfPendingEvents >= maxBatchSize` and by the `flushTimeoutMilliseconds` at the same time?

Comment (DennisPersson) [reply src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:74]: Perhaps you can replace the `BlockingQueue<String> uploadQueue` with a `BlockingQueue<EventBatch> uploadQueue`, and then the UploadPoller thread can just take batches from that? 

So flush() would do something like: 
```
uploadQueue.add(eventStorage.createBatch());
```

in both the timer triggered case and the size triggered case.

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java [hunk 2]
```diff
@@ -44,13 +52,34 @@ public void run() {
...
+          scheduleFlush();
         }
       }
     }
   }
 
+  private void scheduleFlush() {
+    // Cancel the existing scheduled task if it exists
+    if (pendingFlush != null && !pendingFlush.isDone()) {
+      pendingFlush.cancel(false);
```
Comment (DennisPersson) [inline src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:66]: Don't think these `cancel(false)` will do anything:
Screenshot 2024-04-08 at 11 40 43

But if you are making the flush thread safe I don't think it's should be needed?

**Comment by DennisPersson** [comment]
Not sure if this is good inspiration, but this is how the assign logging works in the flag resolver sidecar: https://ghe.spotify.net/konfidens/epx-flags-resolver/blob/master/epx-flags-resolver-sidecar/src/main/java/com/spotify/confidence/flags/resolver/sidecar/AssignLogger.java#L86

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java (hunk 1 above)
Comment (andreas-karlsson) [reply src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:74]: Like we noticed in call @nicklasl createBatch is synchronous, which means there is no concurrency issue. It does however mean we might briefly block the callers thread (sending event).. I think we can live with that for now and create a followup PR where we revisit the queuing strategy, and maybe do what @DennisPersson is suggesting.

**Review by andreas-karlsson — APPROVED** [review]

**Pull request merged**